### ISSUE TITLE
detach_() does not perform in-place detachment from the computation graph

### ISSUE THREAD

**Issue by pfrendl** [pytorch/pytorch#49236]
## 🐛 Bug

detach_() does not perform in-place detachment from the computation graph for a node that already has children. If a gradient computation is issued after the detachment, gradients are still propagated through the node.

## To Reproduce

```
import torch

a = torch.zeros((2, 2), requires_grad=True)
b = 2 * a
c = b.sum()
b.detach_()
c.backward()
print(a.grad)
```

## Expected behavior

Gradients should not be propagated through the detached node, and the past parents should not be referenced anymore, making them collectable by the garbage collector if no further references exist to them.

With the current behavior, it is not possible to perform truncated BPTT with a fixed truncation length without performing redundant computations:
![image](https://user-images.githubusercontent.com/6968154/101929861-aceaab80-3bd7-11eb-9f5c-405ba277cd43.png)
This is especially important in cases where the loss at time t also depends on the model output at time t + c. The sensible thing in such a case would be to have a computation graph at least "c timesteps long". This would be possible if in-place detachment worked as stated above, by detaching the model output at timestep t - c and calling loss.backward(retain_graph=True). I am not aware of any other practical solution.

## Environment

 - PyTorch Version (e.g., 1.0): 1.7.0
 - OS (e.g., Linux): Windows 10
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Python version: 3.8.3
 - CUDA/cuDNN version: 11.0
 - GPU models and configuration: GTX 1080

**Comment by pfrendl** [comment]
The abovementioned use-case is flawed because updating the model parameters after backpropagating through group A makes the source tensors of group B stale, making a backprop through B invalid.
